### DISCUSSION TITLE
mongodb fetch map fail

### DISCUSSION THREAD

**Discussion by chris-yooo** [neuefische/questions#52] [Web] [answered]
hey, ich hab ein problem das sich schwirig zu umschreiben lässt

ich versuche die messages aus der mongo db in ein useState zu schieben, der useState ist zusätzlich auch noch mit fakeMessages initialisiert
![Screenshot (1)](https://user-images.githubusercontent.com/55069641/183610215-944bd389-19ad-4798-ba26-b0a406005ad4.png)

der fetch der nachrichten klappt, und die fakemessages sind auch da, aber er meint das immer beim mappen das message.map keine function ist.

**Comment by PReichetanz** [comment]
Hallo Chris,

wie können wir das Problem reproduzieren, um eine Antwort zu finden? Hilfreich wären ein Link zum richtigen Branch und alle Informationen, die wir brauchen, um an genau die Stelle zu gelangen, an der Du im Screenshot bist (vgl. das [Beispiel von Mareikes Frage hier](https://github.com/orgs/neuefische/discussions/11))?

**Grundsätzlich:**
Die Fehlermeldung "message.map is not a function" weist darauf hin, dass `message` kein Array ist, auf das die `.map()`-Methode angewandt werden kann (weil die Daten noch nicht aus der Datenbank geladen wurden, bis `.map()` ausgeführt wird). Meist hilft hier das [Optional Chaining](https://javascript.info/optional-chaining#optional-chaining): Dort, wo Du über `messages` mapst, musst Du ein `?`setzen:
```js
{messages?.map((message) => ...}
```

**Reply by chris-yooo** [reply]
hab ich getestet selbiger fehler

repo ist dieses: https://github.com/chris-yooo/capstone-project/tree/mongodb

**Comment by chris-yooo** [comment, marked as answer]
lösung ist dieses hier:

```
export default function App() {
  const [messages, setMessages] = useState();
  const [shouldUpdate, setShouldUpdate] = useState(true);
  useEffect(() => {
    async function fetchMessages() {
      try {
        const fetchedMessages = await getMessages();
        setMessages(fetchedMessages);
      } catch (err) {
        console.log(err);
      } finally {
        setShouldUpdate(false);
      }
    }

    if (shouldUpdate) {
      fetchMessages();
    }
  }, [shouldUpdate]);
```

ich hatte vorher fehler in der async function die lässt sich nur so im useEffect aufrufen :)
